drivers/display/display_mcux_lcdifv3.c

```diff
@@ -0,0 +1,425 @@
+			      const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct mcux_lcdifv3_config *config = dev->config;
+	struct mcux_lcdifv3_data *data = dev->data;
+	LCDIF_Type *base = (LCDIF_Type *)DEVICE_MMIO_NAMED_GET(dev, reg_base);
+	uint32_t h_idx;
+	const uint8_t *src;
+	uint8_t *dst;
+
+	__ASSERT((data->pixel_bytes * desc->pitch * desc->height) <= desc->buf_size,
+		 "Input buffer too small");
+
+	LOG_DBG("W=%d, H=%d @%d,%d", desc->width, desc->height, x, y);
+
+	if ((x == 0) && (y == 0) && (desc->width == config->display_config.panelWidth) &&
+	    (desc->height == config->display_config.panelHeight) && (desc->pitch == desc->width)) {
+		/* We can use the display buffer directly, without copying */
+		LOG_DBG("Setting FB from %p->%p", (void *)data->active_fb, (void *)buf);
+		data->active_fb = buf;
+	} else {
+		/* We must use partial framebuffer copy */
+		if (CONFIG_MCUX_LCDIFV3_FB_NUM == 0) {
+			LOG_ERR("Partial display refresh requires driver framebuffers");
+			return -ENOTSUP;
+		} else if (data->active_fb != data->fb[data->next_idx]) {
+			/*
+			 * Copy the entirety of the current framebuffer to new
+			 * buffer, since we are changing the active buffer address
+			 */
+			src = data->active_fb;
+			dst = data->fb[data->next_idx];
+			memcpy(dst, src, config->fb_bytes);
+		}
+		/* Write the display update to the active framebuffer */
+		src = buf;
+		dst = data->fb[data->next_idx];
+		dst += data->pixel_bytes * (y * config->display_config.panelWidth + x);
+
+		for (h_idx = 0; h_idx < desc->height; h_idx++) {
+			memcpy(dst, src, data->pixel_bytes * desc->width);
+			src += data->pixel_bytes * desc->pitch;
+			dst += data->pixel_bytes * config->display_config.panelWidth;
+		}
+		LOG_DBG("Setting FB from %p->%p", (void *)data->active_fb,
+			(void *)data->fb[data->next_idx]);
+		/* Set new active framebuffer */
+		data->active_fb = data->fb[data->next_idx];
+	}
+
+	sys_cache_data_flush_and_invd_range((void *)data->active_fb, config->fb_bytes);
+
+	k_sem_reset(&data->sem);
+
+	/* Set new framebuffer */
+	LCDIFV3_SetLayerBufferAddr(base, 0, (uint32_t)(uintptr_t)data->active_fb);
+	LCDIFV3_TriggerLayerShadowLoad(base, 0);
+
+#if CONFIG_MCUX_LCDIFV3_FB_NUM != 0
+	/* Update index of active framebuffer */
+	data->next_idx = (data->next_idx + 1) % CONFIG_MCUX_LCDIFV3_FB_NUM;
+#endif
+	/* Wait for frame to complete */
+	k_sem_take(&data->sem, K_FOREVER);
+
+	return 0;
+}
+
+static int mcux_lcdifv3_read(const struct device *dev, const uint16_t x, const uint16_t y,
+			     const struct display_buffer_descriptor *desc, void *buf)
+{
+	LOG_ERR("Read not implemented");
+	return -ENOTSUP;
+}
+
+static void *mcux_lcdifv3_get_framebuffer(const struct device *dev)
+{
+	struct mcux_lcdifv3_data *dev_data = dev->data;
+
+	return (void *)dev_data->active_fb;
+}
+
+static int mcux_lcdifv3_display_blanking_off(const struct device *dev)
+{
+	LOG_WRN("Blanking off not implemented");
+	return 0;
```

Comment:
Should these maybe return `-ENOSYS` for unsupported features?

It seems so: https://github.com/zephyrproject-rtos/zephyr/blob/bcc4689cd01620c1a5408986085343e656ec5a71/include/zephyr/drivers/display.h#L400